pandas/core/indexes/period.py

```diff
@@ -707,6 +708,30 @@ def _maybe_convert_timedelta(self, other):
         msg = "Input has different freq from PeriodIndex(freq={0})"
         raise IncompatibleFrequency(msg.format(self.freqstr))
 
+    def _add_offset(self, other):
+        assert not isinstance(other, Tick)
+        base = frequencies.get_base_alias(other.rule_code)
+        if base != self.freq.rule_code:
```

Comment:
you added lots of things here, but don't appear to be tested

Reply:
These are not new cases; they are currently handled in a jumble in maybe_convert_timedeltalike.  This method already exists for DatetimeIndex.

Reply:
Would this be easier to push through if `_add_offset` were separated from the rest of the PR?  This doesn't change any behavior, but does let us remove a particularly inelegant piece of code.

Reply:
well, you are adding code here, not removing code, so not sure what is changing

Reply:
If this is a deal-breaker I can change this to just alias `_add_offset = _add_delta` (which is the method currently called when a `DateOffset` is added to a `PeriodIndex`).  The point of this part of the PR is a) by separating the methods for Tick vs non-Tick offsets, we get to simplify elsewhere, and b) `_add_delta` is a thin wrapper around `_maybe_convert_timedelta`, which is on the messy side.